### PR TITLE
GC fundamentals clean up

### LINKED ISSUE #12783: To improve performance ,disable concurrent GC
In document [ fundamentals of garbage collection ](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals )

> To improve performance when several processes are running, disable concurrent garbage collection

I think concurrent GC is help to improve application performance.It reduce the pauses time when GC is running. Why disable concurrent garbage can improve performance?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: d8948cf5-0aec-308c-8046-965cc15763b2
* Version Independent ID: a30a0be5-b741-2a9c-e997-55d6d9960ca9
* Content: [Fundamentals of garbage collection](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/fundamentals)
* Content Source: [docs/standard/garbage-collection/fundamentals.md](https://github.com/dotnet/docs/blob/master/docs/standard/garbage-collection/fundamentals.md)
* Product: **dotnet**
* Technology: **dotnet-standard**
* GitHub Login: @rpetrusha
* Microsoft Alias: **ronpet**

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #12783

- Add a note that concurrent GC is for 3.5 and earlier and move this section down to bottom
- Remove back to top anchors
- Style improvements

### CODE REVIEW

**Comment by mairaw** [comment]
There are some merge conflicts here

File: docs/standard/garbage-collection/fundamentals.md [hunk 1]
```diff
@@ -50,35 +48,27 @@ The following list summarizes important CLR memory concepts.
...
 
 There is a managed heap for each managed process. All threads in the process allocate memory for objects on the same heap.
 
-To reserve memory, the garbage collector calls the Win32 [VirtualAlloc](/windows/desktop/api/memoryapi/nf-memoryapi-virtualalloc) function, and reserves one segment of memory at a time for managed applications. The garbage collector also reserves segments as needed, and releases segments back to the operating system (after clearing them of any objects) by calling the Win32 [VirtualFree](/windows/desktop/api/memoryapi/nf-memoryapi-virtualfree) function.
+To reserve memory, the garbage collector calls the Win32 [VirtualAlloc](/windows/desktop/api/memoryapi/nf-memoryapi-virtualalloc) function and reserves one segment of memory at a time for managed applications. The garbage collector also reserves segments, as needed, and releases segments back to the operating system (after clearing them of any objects) by calling the Win32 [VirtualFree](/windows/desktop/api/memoryapi/nf-memoryapi-virtualfree) function.
```
Comment (NextTurn) [inline docs/standard/garbage-collection/fundamentals.md]:
```suggestion
To reserve memory, the garbage collector calls the Windows [VirtualAlloc](/windows/desktop/api/memoryapi/nf-memoryapi-virtualalloc) function and reserves one segment of memory at a time for managed applications. The garbage collector also reserves segments, as needed, and releases segments back to the operating system (after clearing them of any objects) by calling the Windows [VirtualFree](/windows/desktop/api/memoryapi/nf-memoryapi-virtualfree) function.
```

Same as line 39.

File: docs/standard/garbage-collection/fundamentals.md [hunk 2]
```diff
@@ -224,61 +214,50 @@ The following are threading and performance considerations for server garbage co
...
 
-- Server garbage collection can be resource-intensive. For example, if you have 12 processes running on a computer that has 4 processors, there will be 48 dedicated garbage collection threads if they are all using server garbage collection. In a high memory load situation, if all the processes start doing garbage collection, the garbage collector will have 48 threads to schedule.
+- Server garbage collection can be resource-intensive. For example, if you have 12 processes running on a computer that has 4 processors, there will be 48 dedicated garbage collection threads if they are all using server garbage collection. In a high memory-load situation, if all the processes start doing garbage collection, the garbage collector has 48 threads to schedule.
 
 If you are running hundreds of instances of an application, consider using workstation garbage collection with concurrent garbage collection disabled. This will result in less context switching, which can improve performance.
 
-<a name="concurrent_garbage_collection"></a>
-
-## Concurrent garbage collection
```
Comment (gewarren) [inline docs/standard/garbage-collection/fundamentals.md]: I moved this section to the end.

File: docs/standard/garbage-collection/fundamentals.md [hunk 3]
```diff
@@ -160,61 +150,45 @@ Before a garbage collection starts, all managed threads are suspended except for
...
-- Server garbage collection, which is intended for server applications that need high throughput and scalability. Server garbage collection can be non-concurrent or background.
+  - In .NET Core, server garbage collection can be non-concurrent or background.
 
-The following illustration shows the dedicated threads that perform the garbage collection on a server.
+  - In .NET Framework 4.5 and later versions, server garbage collection can be non-concurrent or background (background garbage collection replaces concurrent garbage collection). In .NET Framework 4 and previous versions, server garbage collection can be concurrent or non-concurrent.
 
-![Server Garbage Collection Threads](../../../docs/standard/garbage-collection/media/gc-server.png "Server Garbage Collection Threads")
+The following illustration shows the dedicated threads that perform the garbage collection on a server:
 
-### Configuring garbage collection
```
Comment (gewarren) [inline docs/standard/garbage-collection/fundamentals.md]: Moved this section to the end.

**Comment by mairaw** [comment]
There are still merge conflicts here. Do you want to take care of that before we start reviewing?

File: docs/standard/garbage-collection/fundamentals.md [hunk 4]
```diff
@@ -1,42 +1,42 @@
...
   - "garbage collection, managed heap"
 ms.assetid: 67c5a20d-1be1-4ea7-8a9a-92b0b08658d2
 ---
 # Fundamentals of garbage collection
 
-In the common language runtime (CLR), the garbage collector serves as an automatic memory manager. It provides the following benefits:
+In the common language runtime (CLR), the garbage collector (GC) serves as an automatic memory manager. It provides the following benefits:
 
-- Enables you to develop your application without having to manually free memory for objects you create.
+- Enables you to develop your application without having to free memory.
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >without having to free memory. [](start = 42, length = 30)

I think the original wording is better. it emphasizes that you don't need to do this manually. you cannot not free memory - it's just a matter of whether you do it yourself or GC does it for you.

File: docs/standard/garbage-collection/fundamentals.md [hunk 5]
```diff
@@ -136,11 +136,11 @@ A garbage collection has the following phases:
 
   Because generation 2 collections can occupy multiple segments, objects that are promoted into generation 2 can be moved into an older segment. Both generation 1 and generation 2 survivors can be moved to a different segment, because they are promoted to generation 2.
 
-  Ordinarily, the large object heap is not compacted, because copying large objects imposes a performance penalty. However, starting with the .NET Framework 4.5.1, you can use the <xref:System.Runtime.GCSettings.LargeObjectHeapCompactionMode%2A?displayProperty=nameWithType> property to compact the large object heap on demand.
+  Ordinarily, the large object heap is not compacted, because copying large objects imposes a performance penalty. However, in .NET Core and in .NET Framework starting with .NET Framework 4.5.1, you can use the <xref:System.Runtime.GCSettings.LargeObjectHeapCompactionMode%2A?displayProperty=nameWithType> property to compact the large object heap on demand.
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >property to compact the large object heap on demand [](start = 306, length = 51)

in core 3.0, LOH is automatically compacted when running with a hardlimit either by specifying a memory limit on a container or the configs. explained in [https://github.com/dotnet/coreclr/pull/22180](https://github.com/dotnet/coreclr/pull/22180)

File: docs/standard/garbage-collection/fundamentals.md [hunk 6]
```diff
@@ -150,57 +150,45 @@ Before a garbage collection starts, all managed threads are suspended except for
...
 
-You can use the [\<gcServer> element](../../../docs/framework/configure-apps/file-schema/runtime/gcserver-element.md) of the runtime configuration schema to specify the type of garbage collection you want the CLR to perform. When this element's `enabled` attribute is set to `false` (the default), the CLR performs workstation garbage collection. When you set the `enabled` attribute to `true`, the CLR performs server garbage collection.
+  - In .NET Core, server garbage collection can be non-concurrent or background.
 
-Concurrent garbage collection is specified with the [\<gcConcurrent> element](../../../docs/framework/configure-apps/file-schema/runtime/gcconcurrent-element.md) of the runtime configuration schema. The default setting is `enabled`. This setting controls both concurrent and background garbage collection.
+  - In .NET Framework 4.5 and later versions, server garbage collection can be non-concurrent or background (background garbage collection replaces concurrent garbage collection). In .NET Framework 4 and previous versions, server garbage collection can be concurrent or non-concurrent.
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >server garbage collection can be concurrent or non-concurrent [](start = 223, length = 61)

Server GC was never concurrent before 4.5. concurrent Server GC (ie, Background Server GC) was only enabled in 4.5

File: docs/standard/garbage-collection/fundamentals.md [hunk 7]
```diff
@@ -150,57 +150,45 @@ Before a garbage collection starts, all managed threads are suspended except for
...
-- If processor affinity is enabled, you can use the [\<GCHeapAffinitizeMask>](../../framework/configure-apps/file-schema/runtime/gcheapaffinitizemask-element.md) configuration element in .NET Framework applications to control the specific processors for which a GC heap and threads are created. Supply a decimal value. The value is a mask that defines the processors that are available to the process.
-
-### Comparing workstation and server garbage collection
+### Compare workstation and server garbage collection
 
 The following are threading and performance considerations for workstation garbage collection:
 
 - The collection occurs on the user thread that triggered the garbage collection and remains at the same priority. Because user threads typically run at normal priority, the garbage collector (which runs on a normal priority thread) must compete with other threads for CPU time.
 
   Threads that are running native code are not suspended.
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >Threads that are running native code are not suspended. [](start = 2, length = 55)

this is the same for Workstation and Server GC..this is about suspension, not the GC.

File: docs/standard/garbage-collection/fundamentals.md (hunk 4 above)
Comment (BillWagner) [inline docs/standard/garbage-collection/fundamentals.md]: @Maoni0 makes a good point. But I don't like the "objects you create" phrase. It includes objects allocated by the .NET libraries. How about this wording:

```suggestion
- Enables you to develop your application without having to manually free memory.
```

File: docs/standard/garbage-collection/fundamentals.md [hunk 8]
```diff
@@ -1,42 +1,42 @@
...
+  - "garbage collection, concurrent"
+  - "garbage collection, server"
+  - "garbage collection, workstation"
   - "garbage collection, managed heap"
 ms.assetid: 67c5a20d-1be1-4ea7-8a9a-92b0b08658d2
 ---
 # Fundamentals of garbage collection
 
-In the common language runtime (CLR), the garbage collector serves as an automatic memory manager. It provides the following benefits:
+In the common language runtime (CLR), the garbage collector (GC) serves as an automatic memory manager. It provides the following benefits:
```
Comment (BillWagner) [inline docs/standard/garbage-collection/fundamentals.md]: One question I had while reading this article (Hopefully @Maoni0 can answer):  Should I assume that mobile apps built with Xamarin use the Workstation GC? We should state that explicitly.

**Comment by BillWagner** [comment]
@gewarren This really looks good. I reviewed it today, and I like it. I'd be ready to approve it, once you and @Maoni0 are happy with the content.

Let me know when you are ready for a final approval.

File: docs/standard/garbage-collection/fundamentals.md [hunk 9]
```diff
@@ -150,115 +153,125 @@ Before a garbage collection starts, all managed threads are suspended except for
...
 
+For more information about finalization, see <xref:System.Object.Finalize?displayProperty=nameWithType>.
+
 ## Workstation and server garbage collection
 
 The garbage collector is self-tuning and can work in a wide variety of scenarios. You can use a configuration file setting to set the type of garbage collection based on the characteristics of the workload. The CLR provides the following types of garbage collection:
 
-- Workstation garbage collection, which is for all client workstations and stand-alone PCs. This is the default setting for the [\<gcServer> element](../../../docs/framework/configure-apps/file-schema/runtime/gcserver-element.md) in the runtime configuration schema.
+- Workstation garbage collection (GC), which is for all client workstations, stand-alone PCs, and mobile apps built using Xamarin. Workstation GC is the default setting for the [gcServer element](../../../docs/framework/configure-apps/file-schema/runtime/gcserver-element.md) in the run-time configuration schema.
```
Comment (gewarren) [inline docs/standard/garbage-collection/fundamentals.md]: @Maoni0 Please verify statement about mobile apps and workstation gc.

File: docs/standard/garbage-collection/fundamentals.md (hunk 8 above)
Comment (gewarren) [reply docs/standard/garbage-collection/fundamentals.md]: See line 172.

File: docs/standard/garbage-collection/fundamentals.md [hunk 10]
```diff
@@ -150,57 +150,45 @@ Before a garbage collection starts, all managed threads are suspended except for
...
 
 The following are threading and performance considerations for workstation garbage collection:
 
 - The collection occurs on the user thread that triggered the garbage collection and remains at the same priority. Because user threads typically run at normal priority, the garbage collector (which runs on a normal priority thread) must compete with other threads for CPU time.
 
   Threads that are running native code are not suspended.
 
-- Workstation garbage collection is always used on a computer that has only one processor, regardless of the [\<gcServer>](../../../docs/framework/configure-apps/file-schema/runtime/gcserver-element.md) setting. If you specify server garbage collection, the CLR uses workstation garbage collection with concurrency disabled.
+- Workstation garbage collection is always used on a computer that has only one processor, regardless of the [gcServer](../../../docs/framework/configure-apps/file-schema/runtime/gcserver-element.md) setting in .NET Framework apps or the `System.GC.Server` setting in .NET Core apps. If you specify server garbage collection on such a computer, the CLR uses workstation garbage collection with concurrency disabled.
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >If you specify server garbage collection on such a computer, the CLR uses workstation garbage collection with concurrency disabled [](start = 284, length = 130)

this is not true anymore (has never been true on core). we used to do this because there was no concurrent for Server GC.

**Comment by Maoni0** [comment]
> - A heap and a dedicated thread to perform garbage collection are provided for each CPU, and the heaps are collected at the same time. Each heap contains a small object heap and a large object heap, and all heaps can be accessed by user code. Objects on different heaps can refer to each other.

probably a good idea to introduce the SOH/LOH acronyms here

---
Refers to: docs/standard/garbage-collection/fundamentals.md:197 in 3251442. [](commit_id = 325144209fe3e41442918f8cf78282a1ba902d61, deletion_comment = False)

File: docs/standard/garbage-collection/fundamentals.md [hunk 11]
```diff
@@ -210,55 +198,79 @@ The following are threading and performance considerations for server garbage co
...
+- Server garbage collection often has larger size segments. However, this is only a generalization: segment size is implementation-specific and is subject to change. Don't make assumptions about the size of segments allocated by the garbage collector when tuning your app.
 
-- Server garbage collection can be resource-intensive. For example, if you have 12 processes running on a computer that has 4 processors, there will be 48 dedicated garbage collection threads if they are all using server garbage collection. In a high memory load situation, if all the processes start doing garbage collection, the garbage collector will have 48 threads to schedule.
+- Server garbage collection can be resource-intensive. For example, if you have 12 processes running on a computer that has 4 processors, there will be 48 dedicated garbage collection threads if they are all using server garbage collection. In a high memory-load situation, if all the processes start doing garbage collection, the garbage collector has 48 threads to schedule.
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >In a high memory-load situation, if all the processes start doing garbage collection, the garbage collector has 48 threads to schedule. [](start = 241, length = 135)

I would say "if all processes happen to be doing GC at the same time they would interfere with each other as there'd be 12 threads scheduled on the same processor. If these processes are active, it wouldnt be a good idea to have them all use server GC"

File: docs/standard/garbage-collection/fundamentals.md [hunk 12]
```diff
@@ -210,55 +198,79 @@ The following are threading and performance considerations for server garbage co
...
 
-If you are running hundreds of instances of an application, consider using workstation garbage collection with concurrent garbage collection disabled. This will result in less context switching, which can improve performance.
+If you're running hundreds of instances of an application, consider using workstation garbage collection with concurrent garbage collection disabled. This will result in less context switching, which can improve performance.
 
-## Concurrent garbage collection
+## Background garbage collection
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >Background garbage collection [](start = 3, length = 29)

Should say "Background Workstation GC". "Background GC" is ambiguous.

File: docs/standard/garbage-collection/fundamentals.md [hunk 13]
```diff
@@ -210,55 +198,79 @@ The following are threading and performance considerations for server garbage co
...
 
-To improve performance when several processes are running, disable concurrent garbage collection. You can do this by adding a [\<gcConcurrent> element](../../../docs/framework/configure-apps/file-schema/runtime/gcconcurrent-element.md) to the app's configuration file and setting the value of its `enabled` attribute to `"false"`.
+> [!NOTE]
+> Background garbage collection replaces [concurrent garbage collection](#concurrent-garbage-collection) and is available in .NET Framework 4 and later versions. In .NET Framework 4, it's supported only for workstation garbage collection. Starting with .NET Framework 4.5, background garbage collection is available for both workstation and server garbage collection.
 
-Concurrent garbage collection is performed on a dedicated thread. By default, the CLR runs workstation garbage collection with concurrent garbage collection enabled. This is true for single-processor and multi-processor computers.
+A collection on ephemeral generations during background garbage collection is known as foreground garbage collection. When foreground garbage collections occur, all managed threads are suspended.
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: >all managed threads are suspended. [](start = 161, length = 34)

all managed threads are suspended, including the background GC thread which is a managed thread.

File: docs/standard/garbage-collection/fundamentals.md [hunk 14]
```diff
@@ -210,55 +198,79 @@ The following are threading and performance considerations for server garbage co
...
+Concurrent garbage collection enables interactive applications to be more responsive by minimizing pauses for a collection. Managed threads can continue to run most of the time while the concurrent garbage collection thread is running. This results in shorter pauses while a garbage collection is occurring.
 
-The following illustration shows background garbage collection performed on a separate dedicated thread on a server:
+Concurrent garbage collection is performed on a dedicated thread. By default, the CLR runs workstation garbage collection with concurrent garbage collection enabled. This is true for single-processor and multi-processor computers.
+
+The following illustration shows concurrent garbage collection performed on a separate dedicated thread.
+
+![Concurrent Garbage Collection Threads](./media/gc-concurrent.png)
+
+## Configure garbage collection
```
Comment (Maoni0) [inline docs/standard/garbage-collection/fundamentals.md]: > Configure garbage collection [](start = 2, length = 29)

should this just point to the page that you are working on, @gewarren?

File: docs/standard/garbage-collection/fundamentals.md (hunk 9 above)
Comment (Maoni0) [reply docs/standard/garbage-collection/fundamentals.md]: I don't think we should mention Xamarin here because it's a different runtime. 

I would say something like "Workstation garbage collection is designed for client apps and is the default of the GC flavor for stand alone apps. for hosted apps (like by asp.net) the default is decided by the host". 

I'm not sure what "This is the default setting for the gcServer element in the runtime configuration schema." means - should probably just delete it or make it more explicit by say "this means the default value for gcServer in the config is false"

File: docs/standard/garbage-collection/fundamentals.md (hunk 3 above)
Comment (gewarren) [reply docs/standard/garbage-collection/fundamentals.md]: Now just deleted it altogether and link to new topic instead.

File: docs/standard/garbage-collection/fundamentals.md (hunk 14 above)
Comment (gewarren) [reply docs/standard/garbage-collection/fundamentals.md]: Yes, good idea. I added the link, which will be broken until that PR is merged.

**Comment by mairaw** [comment]
I've added a label to indicate this PR is waiting on the other one first.

**Comment by gewarren** [comment]
Close/reopen for new build.

**Comment by gewarren** [comment]
@BillWagner I'm ready for a final approval. Thanks much!

**Review by BillWagner — APPROVED** [review]
This is really well organized, and a great article.

You can :shipit: when ready.

**Comment by gewarren** [comment]
@BillWagner Thanks very much!
